pkg/controllers/resources/persistentvolumeclaims/translate.go

```diff
@@ -15,47 +14,74 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 )
 
-func (s *persistentVolumeClaimSyncer) translate(ctx *synccontext.SyncContext, vPvc *corev1.PersistentVolumeClaim) *corev1.PersistentVolumeClaim {
+var (
+	deprecatedStorageClassAnnotation = "volume.beta.kubernetes.io/storage-class"
+)
+
+func (s *persistentVolumeClaimSyncer) translate(ctx *synccontext.SyncContext, vPvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
 	newPvc := s.TranslateMetadata(vPvc).(*corev1.PersistentVolumeClaim)
-	newPvc = s.translateSelector(ctx, newPvc)
+	newPvc, err := s.translateSelector(ctx, newPvc)
+	if err != nil {
+		return nil, err
+	}
 	if newPvc.Spec.DataSource != nil && vPvc.Annotations[constants.SkipTranslationAnnotation] != "true" &&
 		(newPvc.Spec.DataSource.Kind == "PersistentVolumeClaim" || newPvc.Spec.DataSource.Kind == "VolumeSnapshot") {
-
 		newPvc.Spec.DataSource.Name = translate.PhysicalName(newPvc.Spec.DataSource.Name, vPvc.Namespace)
 	}
 
 	//TODO: add support for the .Spec.DataSourceRef field
-	return newPvc
+	return newPvc, nil
 }
 
-func (s *persistentVolumeClaimSyncer) translateSelector(ctx *synccontext.SyncContext, vPvc *corev1.PersistentVolumeClaim) *corev1.PersistentVolumeClaim {
+func (s *persistentVolumeClaimSyncer) translateSelector(ctx *synccontext.SyncContext, vPvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	vPvc = vPvc.DeepCopy()
+
+	storageClassName := ""
+	if vPvc.Spec.StorageClassName != nil && *vPvc.Spec.StorageClassName != "" {
+		storageClassName = *vPvc.Spec.StorageClassName
+	} else if vPvc.Annotations != nil && vPvc.Annotations[deprecatedStorageClassAnnotation] != "" {
+		storageClassName = vPvc.Annotations[deprecatedStorageClassAnnotation]
+	}
+
+	// translate storage class if we manage those in vcluster
+	if s.storageClassesEnabled {
+		if storageClassName == "" {
+			return nil, fmt.Errorf("no storage class defined for pvc %s/%s", vPvc.Namespace, vPvc.Name)
+		}
```

Comment:
I think this condition should be removed.
Quoting [from docs](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class-1): `PVCs don't necessarily have to request a class. A PVC with its storageClassName set equal to "" is always interpreted to be requesting a PV with no class`